File: 백준/Silver/2164. 카드2/카드2.py
# 파이썬에서는 큐를 스택처럼 list로 활용하며, pop(0)을 사용
# 하지만 list를 사용시 시간초과가 나기 때문에 dequeue 사용
# 길이가 1만 남을 때까지 앞의 수를 버리고, 2번째 수를 뒤로 옮김

import sys
from collections import deque

n = int(sys.stdin.readline())
queue = deque([i for i in range(1,n+1)])
while len(queue) > 1 :
    queue.popleft()
    queue.append(queue.popleft())
print(queue[0])

File: 백준/Silver/2164. 카드2/카드2.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayDeque;
import java.util.Deque;

public class Main {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in)); // 입력
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out)); // 출력
		int N = Integer.parseInt(br.readLine()); // 주어진 수 N 입력
		Deque<Integer> cards = new ArrayDeque<Integer>();
		for(int i=0; i<N; i++) {
			cards.add(i+1);
		}
		for(int i=0; i<N-1; i++) {
			cards.pollFirst();
			cards.add(cards.pollFirst());
		}
		
		
		bw.append(Integer.toString(cards.pop())); // bw에 값 저장
		bw.flush(); // bw 출력
		br.close(); // 입력 종료
		bw.close(); // 출력 종료
	}

}
